codec-http/src/test/java/io/netty/handler/codec/http/HttpServerUpgradeHandlerTest.java

```diff
@@ -0,0 +1,145 @@
+import io.netty.util.CharsetUtil;
+
+import static org.junit.Assert.*;
+
+public class HttpServerUpgradeHandlerTest {
+
+    private class TestUpgradeCodec implements UpgradeCodec {
+        @Override
+        public Collection<CharSequence> requiredUpgradeHeaders() {
+            return Collections.<CharSequence>emptyList();
+        }
+
+        @Override
+        public boolean prepareUpgradeResponse(ChannelHandlerContext ctx, FullHttpRequest upgradeRequest,
+                                              HttpHeaders upgradeHeaders) {
+            return true;
+        }
+
+        @Override
+        public void upgradeTo(ChannelHandlerContext ctx, FullHttpRequest upgradeRequest) {
+            // Ensure that the HttpServerUpgradeHandler is still installed when this is called
+            assertEquals(ctx.pipeline().context(HttpServerUpgradeHandler.class), ctx);
+            assertNotNull(ctx.pipeline().get(HttpServerUpgradeHandler.class));
+
+            // Add a marker handler to signal that the upgrade has happened
+            ctx.pipeline().addAfter(ctx.name(), "marker", new ChannelInboundHandlerAdapter());
+          }
+    }
+
+    @Test
+    public void upgradesPipelineInSameMethodInvocation() {
+        final HttpServerCodec httpServerCodec = new HttpServerCodec();
+        final UpgradeCodecFactory factory = new UpgradeCodecFactory() {
+            @Override
+            public UpgradeCodec newUpgradeCodec(CharSequence protocol) {
+                return new TestUpgradeCodec();
+            }
+        };
+
+        ChannelHandler testInStackFrame = new ChannelDuplexHandler() {
+            // marker boolean to signal that we're in the `channelRead` method
+            private boolean inReadCall;
+            private boolean writeUpgradeMessage;
+            private boolean writeFlushed;
+
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                assertFalse(inReadCall);
+                assertFalse(writeUpgradeMessage);
+
+                inReadCall = true;
+                try {
+                    super.channelRead(ctx, msg);
+                    // All in the same call stack, the upgrade codec should receive the message,
+                    // written the upgrade response, and upgraded the pipeline.
+                    assertTrue(writeUpgradeMessage);
+                    assertFalse(writeFlushed);
+                    assertNull(ctx.pipeline().get(HttpServerCodec.class));
+                    assertNotNull(ctx.pipeline().get("marker"));
+                } finally {
+                    inReadCall = false;
+                }
+            }
+
+            @Override
+            public void write(final ChannelHandlerContext ctx, final Object msg,
+                              final ChannelPromise promise) throws Exception {
+                // We ensure that we're in the read call and defer the write
+                // so we can
```

Comment:
nit: just put this on the previous line. line length of 120 is permitted.

Reply:
The bigger issue was that the comment was incomplete. Sorry about that.